src/algebra/category/CommRing/basic.lean

```diff
@@ -18,103 +18,123 @@ We introduce the bundled categories:
 * `CommSemiRing`
 * `CommRing`
 along with the relevant forgetful functors between them.
+
+## Implementation notes
+
+We make SemiRing (and the other categories) locally reducible in order
+to define its instances. This is because writing, for example,
+
+  instance : concrete_category SemiRing := by { delta SemiRing, apply_instance }
+
+results in an instance of the form `id (bundled_hom.concrete_category _)`
+and this `id`, not being [reducible], prevents a later instance search
+(once SemiRing is no longer reducible) from seeing that the morphisms of
+SemiRing are really semiring morphisms (`→+*`), and therefore have a coercion
+to functions, for example. It's especially important that the `has_coe_to_sort`
+instance not contain an extra `id` as we want the `semiring ↥R` instance to
+also apply to `semiring R.α` (it seems to be impractical to guarantee that
+we always access `R.α` through the coercion rather than directly).
```

Comment:
This explanation is very good and useful. I think it would be nice to have some `example`s that demonstrate that Lean can now see through these syntactic trivialities and find the instances as promised. This comment could include a pointer to those examples (that will likely reside at the bottom of the file). The benefit of such `example`s is that if someone at some point refactors this code, they are forced to take such examples into account.

Reply:
Hmm, I'll think about what could be done here. Other modules (like the ones which construct (co)limits) would break under such a refactoring, but indeed it's clearer if we "test" for the required behavior in isolation.

Reply:
What do we want as examples? I just added

```
example {R S : Mon} (f : R ⟶ S) : (R : Type) → (S : Type) := f
example {R S : CommMon} (f : R ⟶ S) : (R : Type) → (S : Type) := f
```

(in fact, the last one revealed a bug: the coercion from `CommMon` to `Type` was set up incorrectly!)